mesonbuild/mintro.py

```diff
@@ -80,56 +69,58 @@ def list_installed(installdata):
     tlist = []
+
+    # Fast lookup table for installation files
+    intall_lookuptable = {}
+    for i in installdata.targets:
+        outname = os.path.join(installdata.prefix, i.outdir, os.path.basename(i.fname))
+        intall_lookuptable[os.path.basename(i.fname)] = str(pathlib.PurePath(outname))
+
     for (idname, target) in builddata.get_targets().items():
-        t = {'name': target.get_basename(), 'id': idname}
-        fname = target.get_filename()
-        if isinstance(fname, list):
-            fname = [os.path.join(target.subdir, x) for x in fname]
-        else:
-            fname = os.path.join(target.subdir, fname)
-        t['filename'] = fname
-        if isinstance(target, build.Executable):
-            typename = 'executable'
-        elif isinstance(target, build.SharedLibrary):
-            typename = 'shared library'
-        elif isinstance(target, build.StaticLibrary):
-            typename = 'static library'
-        elif isinstance(target, build.CustomTarget):
-            typename = 'custom'
-        elif isinstance(target, build.RunTarget):
-            typename = 'run'
-        else:
-            typename = 'unknown'
-        t['type'] = typename
+        if not isinstance(target, build.Target):
+            raise RuntimeError('Something weird happened. File a bug.')
+
+        fname = [os.path.join(target.subdir, x) for x in target.get_outputs()]
+
+        t = {
+            'name': target.get_basename(),
+            'id': idname,
+            'type': target.get_typename(),
+            'filename': fname,
+            'build_by_default': target.build_by_default,
+            'sources': backend.get_introspection_data(idname, target)
+        }
+
         if installdata and target.should_install():
             t['installed'] = True
-            t['install_filename'] = determine_installed_path(target, installdata)
+            t['install_filename'] = [intall_lookuptable.get(x, None) for x in target.get_outputs()]
         else:
             t['installed'] = False
-        t['build_by_default'] = target.build_by_default
         tlist.append(t)
-    print(json.dumps(tlist))
-
-def list_target_files(target_name, coredata, builddata):
-    try:
-        t = builddata.targets[target_name]
-        sources = t.sources + t.extra_files
-    except KeyError:
-        print("Unknown target %s." % target_name)
-        sys.exit(1)
-    out = []
-    for i in sources:
-        if isinstance(i, mesonlib.File):
-            i = os.path.join(i.subdir, i.fname)
-        out.append(i)
-    print(json.dumps(out))
-
-def list_buildoptions(coredata, builddata):
+    return ('targets', tlist)
+
+def list_target_files(target_name, targets):
+    result = []
+    tgt = None
+
+    for i in targets:
+        if i['id'] == target_name:
+            tgt = i
+            break
+
+    if tgt is None:
+        raise RuntimeError('Target with the ID "{}" could not be found'.format(target_name))
```

Comment:
This changes the return code of meson. It used to be 1 now it's 2. We should not do that.